src/pushover_complete/pushover_api.py

```diff
@@ -50,16 +51,18 @@ def _generic_get(self, endpoint, url_parameter=None, payload=None, session=None)
             raise BadAPIRequestError('{}: {}'.format(resp.status_code, ': '.join(resp_body.get('errors'))))
         return resp_body
 
-    def _generic_post(self, endpoint, url_parameter=None, payload=None, session=None):
+    def _generic_post(self, endpoint, url_parameter=None, payload=None, session=None, files=None):
         """A method for abstracting POST requests to the Pushover API.
 
         :param endpoint: The endpoint of the API to hit. Will be joined with "https://api.pushover.net/1/". Example value: "groups/{}.json"
         :param url_parameter: A parameter to replace in the endpoint string provided. Example value: "g123456". Combined with the above example value, would result in a final URL of "https://api.pushover.net/1/groups/g123456.json"
         :param payload: A dict of parameters to be appended to the URL, e.g. :code:`{'test-param': False}` would result in the URL having :code:`?test-param=false` appended. Do not include the application token in this dict, as it is added by the function.
+        :param files: A dict of file attachment parameters to be attached to the message
         :param session: A :class:`requests.Session` object to be used to send HTTP requests.
         :type endpoint: str
         :type url_parameter: str
         :type payload: dict
+        :type files: dict
```

Comment:
can you specify further what sort of dict this is? More specifically what are they key and value types?

Reply:
Documentation from Pushover is here: https://pushover.net/api#attachments
And I used this page from the Requests documentation: http://docs.python-requests.org/en/latest/user/quickstart/#post-a-multipart-encoded-file

Basically, the dict is of the form:
`{'attachment': (file_name, open(file_path, 'rb'))}`
key must be 'attachment'
value is a tuple of the _name_ of the file (doesn't seem to actually be used by Pushover) and the _file object_
You can also specify the Content-Type of the attachment, but it's not required and made the code simpler where I didn't need to determine what that would be. Though there's probably a very easy way to determine that for a given file.

Reply:
so maybe better `dict[str,tuple[str,Io]]` or whatever the appropriate type is for `Io` since that definitely isn't it. PyCharm introspects on these and gives nice type hints :)

Reply:
Thanks.
`type(open(file_path, 'rb'))` returns `<type 'file'>` so: `dict[str,tuple[str,file]]`?

Reply:
It depends on how requests is implemented - the file class is a much more specific thing that also acts like an IO, so I'm not sure. Maybe look into the requests documentation? This is sort of a nit-picking thing so it's not that big of a deal

Reply:
good point. Another acceptable form of the dictionary is to just pass in a string as the file content instead of using open().
Requests specifies it like this:
```python
    :param files: (optional) Dictionary of ``'name': file-like-objects`` (or ``{'name': file-tuple}``) for multipart encoding upload.
        ``file-tuple`` can be a 2-tuple ``('filename', fileobj)``, 3-tuple ``('filename', fileobj, 'content_type')``
        or a 4-tuple ``('filename', fileobj, 'content_type', custom_headers)``, where ``'content-type'`` is a string
        defining the content type of the given file and ``custom_headers`` a dict-like object containing additional headers
        to add for the file.
```
I suppose I could just copy that.

Reply:
https://github.com/requests/requests/blob/2872f17ab86583294907af84b9c36e63cd9a91ad/requests/api.py#L26

Reply:
I would actually be happy to pass in some sort of iterable of file name(s) (or perhaps just a str/Path and only allow one file?) and leave the creation of the dict and the opening of the file to inside the `_generic_post` function. Forcing the user to pass in a dict with a specific key just seems like extra work for them that could be easily abstracted away by the library.

Reply:
Only a single file can be attached to a given message, so I don't think it makes sense to use an iterable.
Also, the user would: 
1) call send_message(..., image_path, ...) which then 
2) call _send_message(..., image_path, ...); which performs the open, constructs the dict, and then 
3) call _generic_post(..., files_dict)

Doesn't that keep the user from passing in a dict? It seemed like _send_message was the right place to put the dict construction.

Reply:
I've cleaned up the param and type descriptions. It's still a little awkward, if anyone has suggestions for better wording I'm eagerly listening.